Zend/zend_timsort.c

```diff
@@ -0,0 +1,418 @@
+static void *ensureCapacity(struct timsort *ts, size_t minCapacity,
+			    size_t width);
+
+/**
+ * Creates a TimSort instance to maintain the state of an ongoing sort.
+ *
+ * @param a the array to be sorted
+ * @param nel the length of the array
+ * @param c the comparator to determine the order of the sort
+ * @param width the element width
+ */
+static int timsort_init(struct timsort *ts, void *a, size_t len,
+			CMPPARAMS(c, carg),
+			size_t width)
+{
+	int err = 0;
+
+	assert(ts);
+	assert(a || !len);
+	assert(c);
+	assert(width);
+
+	ts->minGallop = MIN_GALLOP;
+	ts->stackSize = 0;
+
+	ts->a = a;
+	ts->a_length = len;
+	ts->c = c;
+#ifdef IS_TIMSORT_R
+	ts->carg = carg;
+#endif
+
+	// Allocate temp storage (which may be increased later if necessary)
+	ts->tmp_length = (len < 2 * INITIAL_TMP_STORAGE_LENGTH ?
+			  len >> 1 : INITIAL_TMP_STORAGE_LENGTH);
+	if (ts->tmp_length) {
+		ts->tmp = malloc(ts->tmp_length * width);
+		err |= ts->tmp == NULL;
+	} else {
+		ts->tmp = NULL;
+	}
+
+	/*
+	 * Allocate runs-to-be-merged stack (which cannot be expanded).  The
+	 * stack length requirements are described in listsort.txt.  The C
+	 * version always uses the same stack length (85), but this was
+	 * measured to be too expensive when sorting "mid-sized" arrays (e.g.,
+	 * 100 elements) in Java.  Therefore, we use smaller (but sufficiently
+	 * large) stack lengths for smaller arrays.  The "magic numbers" in the
+	 * computation below must be changed if MIN_MERGE is decreased.  See
+	 * the MIN_MERGE declaration above for more information.
+	 */
+
+	/* POP:
+	 * In listsort.txt, Tim argues that the run lengths form a decreasing
+	 * sequence, and each run length is greater than the previous two.
+	 * Thus, lower bounds on the minimum runLen numbers on the stack are:
+	 *
+	 *   [      1           = b[1]
+	 *   ,      minRun      = b[2]
+	 *   ,  1 * minRun +  2 = b[3]
+	 *   ,  2 * minRun +  3 = b[4]
+	 *   ,  3 * minRun +  6 = b[5]
+	 *   , ...
+	 *   ],
+	 *
+	 * Moreover, minRun >= MIN_MERGE / 2.  Also, note that the sum of the
+	 * run lenghts is less than or equal to the length of the array.
+	 *
+	 * Let s be the stack length and n be the array length.  If s >= 2, then n >= b[1] + b[2].
+	 * More generally, if s >= m, then n >= b[1] + b[2] + ... + b[m] = B[m].  Conversely, if
+	 * n < B[m], then s < m.
+	 *
+	 * In Haskell, we can compute the bin sizes using the fibonacci numbers
+	 *
+	 *     fibs = 1:1:(zipWith (+) fibs (tail fibs))
+	 *
+	 *     cumSums a = case a of { [] -> [] ; (x:xs) -> x:(map (x+) (cumSums xs)) }
+	 *
+	 *     fibSums = cumSums fibs
+	 *
+	 *     binSizes minRun = ([ 1, minRun, minRun + 2 ]
+	 *                        ++ [ (1 + minRun) * (fibs !! (i+2))
+	 *                             + fibSums !! (i+1) - fibs !! i | i <- [0..] ])
+	 *
+	 *     arraySizes minRun = cumSums (binSizes minRun)
+	 *
+	 * We these funcitons, we can compute a table with minRun = MIN_MERGE / 2 = 16:
```

Comment:
`We these funcitons` -> `With these functions` ?

(the original code has the same "mistake": https://github.com/patperry/timsort/blob/9fc97ae0857e4fbce06c1b6b26b48e427e307895/timsort.c#L262)